pkg/distributor/distributor.go

```diff
@@ -818,12 +819,19 @@ func (d *Distributor) prePushRelabelMiddleware(next push.Func) push.Func {
 			ts := req.Timeseries[tsIdx]
 
 			if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
-				l, keep := relabel.Process(mimirpb.FromLabelAdaptersToLabels(ts.Labels), mrc...)
+				lb := labels.NewBuilder(mimirpb.FromLabelAdaptersToLabels(ts.Labels))
+				lb.Set(model.MetaLabelPrefix+"tenant_id", userID)
+				keep := relabel.ProcessBuilder(lb, mrc...)
 				if !keep {
 					removeTsIndexes = append(removeTsIndexes, tsIdx)
 					continue
 				}
-				ts.Labels = mimirpb.FromLabelsToLabelAdapters(l)
+				lb.Range(func(x labels.Label) {
+					if strings.HasPrefix(x.Name, model.MetaLabelPrefix) {
+						lb.Del(x.Name)
+					}
+				})
+				ts.Labels = mimirpb.FromLabelsToLabelAdapters(lb.Labels(nil))
```

Comment:
We should be using `labels.EmptyLabels()` instead of `nil`, that will work with both "old" and new `stringlabels` labels:

```suggestion
				ts.Labels = mimirpb.FromLabelsToLabelAdapters(lb.Labels(labels.EmptyLabels()))
```

https://prometheus.io/blog/2023/03/21/stringlabel/

[*] Although to be fair, entire `mimirpb.FromLabelsToLabelAdapters` conversion will no longer work anyway and needs fixing.

Reply:
Great news, thanks for the link!